app/utils/LocationFinder.scala

```diff
@@ -0,0 +1,98 @@
+package utils
+
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+import play.api.libs.ws._
+import play.api.Play.current
+
+import scala.concurrent.{Await, Future}
+import scala.concurrent.duration._
+import scala.math._
+import scala.util.{Success, Failure}
+
+object LocationFinder {
+  case class Location(lat: Double, lng: Double)
+  case class Place(name: String, location: Location)
+  case class Places(places: Seq[Place])
+
+
+  // Implicit Reads for case classes
+  implicit val locationReads : Reads[Location] = (
+    (JsPath \ "lat").read[Double] and
+    (JsPath \ "lng").read[Double]
+  )(Location.apply _)
+
+  implicit val placeReads : Reads[Place] = (
+    (JsPath \ "name").read[String] and
+    (JsPath \ "geometry" \ "location").read[Location]
+  )(Place.apply _)
+
+  // Explicit Reads for Places case class
+  val placesReads : Reads[Places] = (
+    (JsPath \ "results").read[Seq[Place]].map(Places(_))
+  )
+
+
+  // Concurrent execution context for Future
+  implicit val context = scala.concurrent.ExecutionContext.Implicits.global
+
+
+  // Find Euclidean distance between two GPS coordinates
+  def dist(a: Location, b: Location): Double = {
+    sqrt(pow(a.lat - b.lat, 2) + pow(a.lng - b.lng, 2))
+  }
+
+  // Find the nearest location to the given current location, given the nearest
+  // key location
+  def nearestLocation(currLoc: Location, 
+                      keyLoc: Option[Place] = None): Option[Place] = {
+    val placeLoc = nearestPlacesLocation(currLoc)
+    var placeDist = Double.MaxValue
+    if (placeLoc isDefined) {
+      placeDist = dist(currLoc, placeLoc.get.location)
+    }
+
+    var keyDist = Double.MaxValue
+    if (keyLoc isDefined) {
+      keyDist = dist(currLoc, keyLoc.get.location)
+    }
+
+    if (keyDist == Double.MaxValue && placeDist == Double.MaxValue) {
+      None
+    } else if (keyDist <= placeDist) {
+      keyLoc
+    } else {
+      placeLoc
+    }
+  }
+
+  def nearestPlacesLocation(currLoc: Location): Option[Place] = {
```

Comment:
This could use some comments to explain what's going on with this Google API request.